File: Python/day_4.py
def solve():
    with open("testing.txt", "r") as f:
        lines = f.readlines()

    total_a = 0
    total_b = 0

    for line in lines:
        line = line.split(",")
        line[0] = line[0].split('-')
        line[1] = line[1][:-1].split('-')
        a, b = int(line[0][0]), int(line[0][1])
        c, d = int(line[1][0]), int(line[1][1])

        if ((b >= d) and (a <= c)) or ((d >= b) and (c <= a)):
            total_a += 1
        if (d >= b >= c) or (c <= a <= d) or (a <= c <= b) or (a <= d <= b):
            total_b += 1

    print(total_a, total_b)


if __name__ == "__main__":
    solve()

File: Rust/src/day_4.rs


pub fn solution() {
    let input = include_str!("../../inputs/day_4.txt")
        .lines()
        .map(|l| l.split_once(',').unwrap())
        .map(|(a, b)| (a.split_once('-').unwrap(),
             b.split_once('-').unwrap()))
        .map(|((c, d), (e, f))| (c.parse::<i32>().unwrap(), d.parse::<i32>().unwrap(),
                                 e.parse::<i32>().unwrap(), f.parse::<i32>().unwrap())).collect::<Vec<(i32, i32, i32, i32)>>();

    //println!("{:?}", input);

    let mut part_a_total = 0;
    let mut part_b_total = 0;

    // if ((b >= d) and (a <= c)) or ((d >= b) and (c <= a)):
    //     total_a += 1
    // if (d >= b >= c) or (c <= a <= d) or (a <= c <= b) or (a <= d <= b):
    //     total_b += 1

    // Part A
    for (a, b, c, d) in input.iter() {
        if ((*a <= *c) & (*b >= *d)) | ((*d >= *b) & (*c <= *a)) {
            part_a_total += 1;
        }
        if ((*d >= *b) & (*b >= *c)) | ((*c <= *a) & (*a <= *d)) | ((*a <= *c) & (*c <= *b)) | ((*a <= *d) & (*d <= *b)) {
            part_b_total += 1;
        }
    }
    println!("Part A total: {}\nPart B total: {}", part_a_total, part_b_total)

}